JOSS Paper (#670)
tqec/tqec: paper/paper.md

  orcid: 0009-0008-7933-6848
  affiliation: 11
- name: Brendan Reid
  affiliation: 12
- name: Austin Fowler
  equal_contrib: true
  affiliation: 13

affiliations:
- name: Qraftware, Toulouse, France
  index: 1
- name: University of Science and Technology of China
  index: 2
- name: School of Physics and Applied Physics, Southern Illinois University, Carbondale, IL, 62901, USA
  index: 3
- name: Institute of Computing Technology, Chinese Academy of Sciences
  index: 4
- name: Department of Computer Science, Northwestern University, United States
  index: 5
- name: Independent Consultant
  index: 6
- name: Independent Researcher
  index: 7
- name: Technical University of Munich, TUM School of Computation, Information and Technology
  index: 8
- name: Solvy
  index: 9
- name: Parity Quantum Computing Germany GmbH, 20095 Hamburg, Germany
  index: 10
- name: VTT, Finland
  index: 11
- name: PsiQuantum
  index: 12
- name: Stairway Invest
  index: 13

date: 27 August 2025
bibliography: paper.bib

---

# Summary

`tqec` is a Python-based open-source compiler that takes a logical-level quantum
computation model represented as connected 3D primitive blocks and translates it into a detailed,

> fault-tolerant, physical-level circuit. The result is a Stim circuit with all the detailed

Reviewer: ```suggestion
fault-tolerant, physical-level circuit. The result is a `Stim` circuit with all the detailed
```
minor nit

designers and experimentalists to rapidly iterate and obtain exact low-level circuits, facilitating
efficient performance simulation or experimental demonstration. The ‘Topological Quantum Error Correction’
of the package name highlights the project’s focus on topological codes. At present, the project is
primarily centered on the surface code.

Simulations of quantum computer operations in the large-scale error correction regime are currently
infeasible. Building the logical Stim circuits is a hassle, and Monte Carlo simulations at the scale of,
for example, hierarchical memory systems involving yoked surface codes, are difficult to perform exactly.
The full-scale simulations performed by `tqec` provide more accurate fault-tolerant resource estimation than empirical extrapolations.

`tqec` is based on several academic papers (@polian2015, @fowler:2012, @mcewen:2023, @gidney:2025, @kissinger:2020),
as well as making substantial use of Craig Gidney's Stim package @gidney:2021.

Some recent work enabled by `tqec`: @kan:2025, @dubey:2025, @bolanos:2025.

# Acknowledgements

We thank the Unitary Foundation for a micro-grant in the early stages of the project development.

> # References

Reviewer: See https://pandoc.org/MANUAL.html#placement-of-the-bibliography for adding references to this section. 

Not too sure if this is helpful

Author: If I'm reading it right, it seems to suggest what I've done is correct, and ought to print a list of the references at the end of the paper.

Reviewer: managed to get the workflow working 🥳 

There are a couple of build warnings about references:

https://github.com/tqec/tqec/actions/runs/17167679496/job/48711453571?pr=670

<img width="642" height="140" alt="image" src="https://github.com/user-attachments/assets/26e2fede-20df-4ce6-8a24-d9be8f72c263" />

<img width="514" height="54" alt="image" src="https://github.com/user-attachments/assets/eb11479a-e460-4dc8-9a7e-9887018f18a6" />

---
title: 'tqec: A Python package for topological quantum error correction'

authors:
- name: Adrien Suau
  equal_contrib: true
  affiliation: 1
- name: Yiming Zhang
  equal_contrib: true
  affiliation: 2
- name: Purva Thakre
  affiliation: 3

> - name: Yilun Zhao

Reviewer: @Zhaoyilunnn can you make a suggested change with your affiliation?

- name: Kabir Dubey
  orcid: 0000-0002-7621-9004
  affiliation: 4
- name: Jose A Bolanos
- name: Arabella Schelpe
- name: Philip Seitz
- name: Gian Giacomo Guerreschi
- name: Ángela Elisa Álvarez Pérez
  affiliation: 5
- name: Reinhard Stahn
- name: Jerome Lensen

> - name: Brendan Reid

Author: Ah I missed that. Thanks! I'll bring them up to date

- name: Austin Fowler
  equal_contrib: true
  affiliation: 6

affiliations:
- name: Qraftware
  index: 1
- name: University of Science and Technology of China
  index: 2
- name: School of Physics and Applied Physics, Southern Illinois University, Carbondale, IL, 62901, USA
  index: 3
- name: Department of Computer Science, Northwestern University, United States
  index: 4
- name: Solvy
  index: 5
- name: Stairway Invest
  index: 6

date: 15 August 2025
bibliography: paper.bib

---

# Summary

> `tqec` is a Python-based open-source project. It is a compiler, which takes a logical-level quantum

Reviewer 2: ```suggestion
`tqec` is a Python-based open-source project. It is a compiler which takes a logical-level quantum
```

computation model represented as connected 3D primitive blocks and translates it into a detailed,
fault-tolerant, physical-level circuit. The result is a Stim circuit with all the detailed
information needed for simulation or to run on real quantum hardware. This enables both quantum algorithm
designers and experimentalists to rapidly iterate and obtain exact low-level circuits, facilitating
efficient performance simulation or experimental demonstration. The ‘Topological Quantum Error Correction’
of the package name highlights the project’s focus on topological codes. At present, the project is
primarily centered on the surface code.

> 
> Simulations of quantum computer operations in the large-scale error correction regime are currently
> infeasible. Building the logical Stim circuits is a hassle, and Monte Carlo simulations at the scale of,
> for example, hierarchical memory systems involving yoked surface codes, are difficult to perform exactly.
> The full-scale simulations performed by`tqec` provide evidence of potential opportunities to reduce
> overhead in quantum circuits at a greater level of detail than extrapolations.
> 
> `tqec` is based on several academic papers (@polian:2015, @fowler:2012, @mcewen:2023, @gidney:2025, @kissinger:2020),
> as well as making substantial use of Craig Gidney's Stim package @gidney:2021.
> 
> Some recent work enabled by `tqec`: @kan:2025, @dubey:2025, @bolanos:2025.

Reviewer 2: Changes to address [this requirement](https://joss.readthedocs.io/en/latest/review_checklist.html):

> A statement of need: Does the paper have a section titled ‘Statement of need’ that clearly states what problems the software is designed to solve, who the target audience is, and its relation to other work?

```suggestion

# Statement of need

Simulations of quantum computer operations in the large-scale error correction regime are currently
infeasible. Building the logical Stim circuits is a hassle, and Monte Carlo simulations at the scale of,
for example, hierarchical memory systems involving yoked surface codes, are difficult to perform exactly @gidney:2025.
The full-scale simulations performed by`tqec` provide evidence of potential opportunities to reduce
overhead in quantum circuits at a greater level of detail than extrapolations.

`tqec` is designed to be used by students and researchers that seek to understand the theory of quantum error correction and experiment with scalable quantum computer system and circuit designs. Posters featuring preliminary research and an educational tutorial enabled by `tqec` have been approved for conference proceedings: @kan:2025, @dubey:2025, @bolanos:2025. The functionality of the package is based on several academic papers (@polian:2015, @fowler:2012, @mcewen:2023, @gidney:2025, @kissinger:2020), and makes substantial use of Craig Gidney's Stim package @gidney:2021.
```

Author: Ah, I didn't realise. That's great. I'll leave those citations as is then.

---
title: 'tqec: A Python package for topological quantum error correction'

authors:
- name: Adrien Suau
  equal_contrib: true
  affiliation: 1
- name: Yiming Zhang
  equal_contrib: true
  affiliation: 2
- name: Purva Thakre
  affiliation: 3

> - name: Yilun Zhao

Reviewer 3: ```suggestion
- name: Yilun Zhao
  orcid: 0000-0002-6812-5120
  affiliation: 11
```

- name: Kabir Dubey
  orcid: 0000-0002-7621-9004
  affiliation: 4
- name: Jose A Bolanos
- name: Arabella Schelpe
- name: Philip Seitz
  orcid: 0000-0003-3856-4090
  affiliation: 5
- name: Gian Giacomo Guerreschi
- name: Ángela Elisa Álvarez Pérez
  affiliation: 6
- name: Reinhard Stahn
  affiliation: 7
- name: Jerome Lensen
  orcid: 0009-0008-7933-6848
  affiliation: 8
- name: Brendan Reid
  affiliation: 9
- name: Austin Fowler
  equal_contrib: true
  affiliation: 10

affiliations:
- name: Qraftware, Toulouse, France
  index: 1
- name: University of Science and Technology of China
  index: 2
- name: School of Physics and Applied Physics, Southern Illinois University, Carbondale, IL, 62901, USA
  index: 3
- name: Department of Computer Science, Northwestern University, United States
  index: 4
- name: Technical University of Munich, TUM School of Computation, Information and Technology
  index: 5
- name: Solvy
  index: 6
- name: Parity Quantum Computing Germany GmbH, 20095 Hamburg, Germany
  index: 7
- name: VTT, Finland
  index: 8
- name: PsiQuantum
  index: 9
- name: Stairway Invest
  index: 10

> 

Reviewer 3: ```suggestion
- name: Institute of Computing Technology, Chinese Academy of Sciences
  index: 11
```

fault-tolerant, physical-level circuit. The result is a Stim circuit with all the detailed
information needed for simulation or to run on real quantum hardware. This enables both quantum algorithm
designers and experimentalists to rapidly iterate and obtain exact low-level circuits, facilitating
efficient performance simulation or experimental demonstration. The ‘Topological Quantum Error Correction’
of the package name highlights the project’s focus on topological codes. At present, the project is
primarily centered on the surface code.

# Statement of Need

Simulations of quantum computer operations in the large-scale error correction regime are currently
infeasible. Building the logical Stim circuits is a hassle, and Monte Carlo simulations at the scale of,
for example, hierarchical memory systems involving yoked surface codes, are difficult to perform exactly, @gidney:2025.
The full-scale simulations performed by `tqec` provide more accurate fault-tolerant resource estimation
than empirical extrapolations.

`tqec` is designed to be used by students and researchers who seek to understand the theory of quantum

> error correction and experiment with scalable quantum computer system and circuit designs. Posters featuring

Reviewer: ```suggestion
error correction and experiment with scalable quantum computer systems and circuit designs. Posters featuring
```

Author: I agree with the second change - I should have caught that in changing the topologiq reference, thanks! I'm not sure about the first change. This was @KabirDubey's sentence. I had read it as a shortening of "quantum computer system designs and circuit designs", rather than two separate phrases with an "and" in between. Maybe @KabirDubey could confirm which he meant?

Author: thanks for confirming!

preliminary research and an educational tutorial enabled by `tqec` have been approved for conference

> proceedings: @kan:2025 and @dubey:2025. A further software package has been recently built to enable better interfacing

Reviewer: ```suggestion
proceedings: @dubey:2025 and @kan:2025 . A further software package has been recently built to enable better interfacing
```
